### PR TITLE
fix(zebrad): retry explicit missing inventory without poisoning peers

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

High-concurrency block sync could fail to recover cleanly after missing predecessor blocks. `MissingInventoryCollector::send()` marked all inventory in a failed request as missing for any request error, including timeouts, dropped connections, and overload. That collapsed "request failed" into "peer does not have this block."

Once the registry had those false misses, later downloads could fail with `PeerError::NotFoundRegistry` wrapped in `SharedPeerError` and `BlockDownloadVerifyError::DownloadFailed`. The sync loop then had too little typed information: peer `notfound`, registry `notfound`, duplicate commit errors, and generic transport errors were partly distinguished by debug-string checks. In local Z3 testing with high concurrency, this showed up as sync repeatedly failing or restarting instead of prioritizing the missing predecessor and continuing.

The conclusion came from Z3 mainnet and testnet observations plus code path tracing. Logs and metrics showed sync progress and restart behavior; code inspection showed all request errors updated the missing-inventory registry; the sync loop handled `notfound` without a bounded predecessor retry. Existing issue #2908 also captures the need for typed `VerifyBlockError::Commit` handling so sync decisions do not depend on fragile error wrapping.

The failure mode is not specific to tip operation; IBD nodes hit the same cycle. With `checkpoint_verify_concurrency_limit` near its default of 1000, the syncer's lookahead saturates with hundreds of in-flight blocks. When `BLOCK_VERIFY_TIMEOUT` trips on a stuck predecessor, the registry poisoning takes out the whole batch. The next sync attempt then logs `exhausted prospective tip set`.

Related to #10630 

## Solution

This keeps the fix focused on the broken contract: missing inventory is now a typed, explicit signal, not a side effect of any transport failure.

- Keep missing-inventory registry updates to explicit `notfound` responses. Generic transport failures no longer mark all requested inventory as missing.
- Preserve typed `PeerErrorKind` and explicit missing inventory in `SharedPeerError`, so sync can classify peer `notfound` and local registry `notfound` without debug-string matching.
- Add a bounded in-place retry for peer-reported missing blocks. When a predecessor is missing, Zebra cancels newer in-flight downloads, retries the missing hash first, then requeues interrupted work behind it.
- Restart sync when the local registry says all ready peers are missing a block, or when the per-hash retry limit is exhausted.
- Use zero delay only after successful progress, and a short 5-second delay when a failed sync run still made progress. This keeps responsive catch-up without a tight loop.
- Convert wrapped `CommitSemanticallyVerifiedError` duplicate commits back into `VerifyBlockError::Commit`, so duplicate classification remains typed through the semantic verifier.
- Remove dead request inventory helpers and add targeted tests around missing-inventory retry, restart classification, single-hash tip behavior, and duplicate commit wrapping.

### Tests

- Local Z3 stack deployment on mainnet and testnet:
  - Post-deploy logs showed mainnet accepting blocks `3355579` and `3355580`, and testnet accepting blocks `4031310` through `4031316`.
  - Recent logs showed no panic, fatal error, OOM, `notfound`, or missing-inventory retry loop.

### Specifications & References

- Related: #2908

### Follow-up Work

No immediate follow-up is required for this PR. Broader sync architecture work can separately evaluate peer scoring or stricter interrupted-hash ordering, but neither is required to fix the registry poisoning and bounded missing-inventory recovery path.

### AI Disclosure

- [ ] No AI tools were used in this PR
- [x] AI tools were used: OpenAI coding assistant for implementation support, review cleanup, PR drafting, and validation orchestration.

### PR Checklist

- [x] The PR title follows [conventional commits](https://www.conventionalcommits.org/) format: `type(scope): description`
- [x] The PR follows the [contribution guidelines](https://github.com/ZcashFoundation/zebra/blob/main/CONTRIBUTING.md).
- [x] This change was discussed in an issue or with the team beforehand.
- [x] The solution is tested.
- [x] The documentation and changelogs are up to date.

### CODE REVIEW

File: zebra-network/src/peer/error.rs [hunk 1]
```diff
@@ -9,29 +9,104 @@ use zebra_chain::serialization::SerializationError;
...
+
+impl fmt::Display for SharedPeerError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(self.source.as_ref(), f)
+    }
+}
+
+impl std::error::Error for SharedPeerError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        std::error::Error::source(self.source.as_ref())
```
[the remark this answers is not in the document]
Comment (gustavovalverde) [reply zebra-network/src/peer/error.rs:51]: I checked the `tracing-error` implementation and kept the delegation intentional: `TracedError::source()` exposes the erased span-trace source used by `ExtractSpanTrace`. Returning `self.source` would add a new `TracedError` layer rather than preserving the existing source-chain shape. I added a short code comment to make that explicit.

**Review by oxarbitrage — COMMENTED** [review]
Is the retry path necessary to observe the fix working in practice, or is it optional for this PR?

**Comment by gustavovalverde** [comment]
> Is the retry path necessary to observe the fix working in practice, or is it optional for this PR?

It is necessary for the practical fix, yes.

**Comment by oxarbitrage** [comment]
Closing in favor of #10442, which addresses the root cause at the peer routing level — block requests are only sent to peers at or above the current sync height, so the NotFoundRegistry is no longer polluted by stalled peers returning genuine `notfound`. The false-miss bug (#10633) is still present in the code but no longer triggers in practice. Can be reopened as a hardening fix post-merge if desired.

**Comment by gustavovalverde** [comment]
@oxarbitrage I agree that #10442 fixes **one important** trigger path: multi-hash block downloads are routed to peers whose handshake height is at or above Zebra's current chain tip. That should reduce requests sent to peers that are obviously behind, and therefore reduce genuine `notfound` responses from lagging peers.

But I do **not** agree that it makes this PR unnecessary or that the false-miss bug "no longer triggers in practice" as an architectural conclusion, because #10442:

1. Does not remove the false-miss behavior: `MissingInventoryCollector::send` is unchanged. It still treats generic inventory request errors, including timeouts and dropped connections, as missing inventory.

2. Does not route **all** block requests through the height filter: It routes **multi-hash** `BlocksByHash` through the new block-download path. Single-hash `BlocksByHash` still goes through inventory-aware `route_inv`, and the new route also falls back to normal P2C if no peers meet the height threshold.

3. The peer height is handshake `start_height`, not a live peer tip: It is useful, but it is stale by design. It is a filter, not proof the peer currently has the requested block.

4. The batching path hides the original failure cause: It turns failed batch downloads into `DownloadFailed { hash }`, so it loses whether the failure was `notfound`, `NotFoundRegistry`, timeout, connection closed, overload, etc. That makes the system less able to distinguish root causes.
